File: frontend/src/components/benchmark/content/jaccard_similarity/implementations/jaccard_similarity.js
class JaccardSimilarity {
    constructor() {
        this.adjacency = new Map();
        this.nodes = new Set();
    }

    addEdge(u, v) {
        if (!this.adjacency.has(u)) {
            this.adjacency.set(u, new Set());
        }
        if (!this.adjacency.has(v)) {
            this.adjacency.set(v, new Set());
        }
        this.adjacency.get(u).add(v);
        this.adjacency.get(v).add(u);
        this.nodes.add(u);
        this.nodes.add(v);
    }

    buildFromEdges(edges) {
        for (const [u, v] of edges) {
            this.addEdge(u, v);
        }
    }

    jaccardSimilarity(u, v) {
        if (!this.adjacency.has(u) || !this.adjacency.has(v)) {
            return 0.0;
        }

        const neighborsU = this.adjacency.get(u);
        const neighborsV = this.adjacency.get(v);

        let intersection = 0;
        for (const node of neighborsU) {
            if (neighborsV.has(node)) {
                intersection++;
            }
        }

        const unionSize = neighborsU.size + neighborsV.size - intersection;
        if (unionSize === 0) {
            return 0.0;
        }

        return intersection / unionSize;
    }

    scoreAllPairs() {
        const scores = [];
        const nodeList = Array.from(this.nodes);

        for (let i = 0; i < nodeList.length; i++) {
            for (let j = i + 1; j < nodeList.length; j++) {
                const u = nodeList[i];
                const v = nodeList[j];
                const score = this.jaccardSimilarity(u, v);
                if (score > 0) {
                    scores.push({ pair: [u, v], score: score });
                }
            }
        }

        scores.sort((a, b) => b.score - a.score);
        return scores;
    }

    topKPredictions(k = 10) {
        const scores = this.scoreAllPairs();
        return scores.slice(0, k);
    }
}

module.exports = JaccardSimilarity;

File: frontend/src/components/benchmark/content/jaccard_similarity/implementations/jaccard_similarity.go
package main

import (
    "sort"
)

type JaccardSimilarity struct {
    adjacency map[int]map[int]bool
    nodes     map[int]bool
}

func NewJaccardSimilarity() *JaccardSimilarity {
    return &JaccardSimilarity{
        adjacency: make(map[int]map[int]bool),
        nodes:     make(map[int]bool),
    }
}

func (js *JaccardSimilarity) AddEdge(u, v int) {
    if js.adjacency[u] == nil {
        js.adjacency[u] = make(map[int]bool)
    }
    if js.adjacency[v] == nil {
        js.adjacency[v] = make(map[int]bool)
    }
    js.adjacency[u][v] = true
    js.adjacency[v][u] = true
    js.nodes[u] = true
    js.nodes[v] = true
}

func (js *JaccardSimilarity) BuildFromEdges(edges [][2]int) {
    for _, edge := range edges {
        js.AddEdge(edge[0], edge[1])
    }
}

func (js *JaccardSimilarity) JaccardSimilarity(u, v int) float64 {
    if js.adjacency[u] == nil || js.adjacency[v] == nil {
        return 0.0
    }

    neighborsU := js.adjacency[u]
    neighborsV := js.adjacency[v]

    intersection := 0
    for node := range neighborsU {
        if neighborsV[node] {
            intersection++
        }
    }

    unionSize := len(neighborsU) + len(neighborsV) - intersection
    if unionSize == 0 {
        return 0.0
    }

    return float64(intersection) / float64(unionSize)
}

func (js *JaccardSimilarity) ScoreAllPairs() [][3]interface{} {
    var scores [][3]interface{}
    var nodeList []int
    for node := range js.nodes {
        nodeList = append(nodeList, node)
    }

    for i := 0; i < len(nodeList); i++ {
        for j := i + 1; j < len(nodeList); j++ {
            u := nodeList[i]
            v := nodeList[j]
            score := js.JaccardSimilarity(u, v)
            if score > 0 {
                scores = append(scores, [3]interface{}{u, v, score})
            }
        }
    }

    sort.Slice(scores, func(i, j int) bool {
        return scores[i][2].(float64) > scores[j][2].(float64)
    })

    return scores
}

func (js *JaccardSimilarity) TopKPredictions(k int) [][3]interface{} {
    scores := js.ScoreAllPairs()
    if k > len(scores) {
        k = len(scores)
    }
    return scores[:k]
}